### PR TITLE
Fix distributed tripolar fold

### PR DESCRIPTION
**Pull request by briochemc**
This PR fixes the distributed tripolar fold with no extra MPI passes, for both UPivot and FPivot topologies. The current distributed UPivot on main is broken (halos don't match the serial case), so this PR addresses that and adds FPivot support.

### New distributed fold topologies

Doing this without extra MPI passes requires local y-topologies that encode more information than just `FullyConnected`. This is accomplished by introducing 4 new topology types for distributed tripolar grid fold boundaries:
- `LeftConnectedRightCenterFolded` (1xN slab UPivot)
- `LeftConnectedRightFaceFolded` (1xN slab FPivot)
- `LeftConnectedRightCenterConnected` (MxN pencil UPivot)
- `LeftConnectedRightFaceConnected` (MxN pencil FPivot)

These replace the previous `FullyConnected` y-topology on northernmost distributed ranks.

### Fold-aware MPI communication buffers

Three buffer types encode fold geometry for the distributed zipper for pencil partitions (slab partitions use the serial fold BC directly):
- `TwoDZipperBuffer` — interior-width x-reversal with `FL`/`WFL` flags controlling fold-line inclusion and write ownership based on pivot position.
- `ZipperCornerBuffer` — NW/NE corner halos where the fold intersects x-boundaries between ranks.
- `TripolarXBuffer` — X-direction halos adjacent to the fold with matching fold-line awareness.

### Worksize-based kernel coverage for FPivot

Added `worksize` override for distributed FPivot grids returning `(Nx, Ny+1, Nz)`, matching the serial override from #5408. Extended the async distributed buffer tendency kernel parameters to also use `worksize(grid)` instead of `size(grid)`, closing a coverage gap at the fold line that caused serial/distributed divergence after 2 time steps.

## TODO: extend testing

I have a bunch of extra tests that I have run locally, and which seem to pass. I intend to include some here, in particular to cover the FPivot topology. Future commits could add to CI:
- FPivot versions of existing distributed tripolar tests (grid/field reconstruction, boundary conditions, simulations)
- Index-tracing halo verification for all field locations × partitions × fold topologies
- Per-rank local field + halo comparison against serial parent array. This is a bit like the index-tracing vizualizations for each rank (see example below).

## Pivot point mismatch

There is one single point where serial and distributed halo-filling don't match: the pivot points. That's because it is a very special case that would need special handling, and I don't think we should "waste" lines of codes for it since it will always be covered by land. Take the ranks A and B "touching" the central pivot point pictured below (for a FPivot pencil partition; the star represents the pivot point, which is at FF location for this topology). When you fill the halos of A around the pivot, you use the north, corner, and x buffers. The buffers all have nice "rectangle like" shapes. But the pivot point must come from B, despite being on a row coming from A itself through the corner buffer. And this only happens at the pivot, so for any partition (Rx,Ry) where Rx>2 it would need an extra type/encoding to correctly fill that single point only for this rank next to the pivot. So I think it's not worth it, but happy to be convinced otherwise.

IMG_233B6A9D9E52-1

## Visual verification

The fold is verified visually using index-tracking fields with half-circles representing 180° rotation around pivot points. For a 4x2 UPivot 20x20 grid with `Hx=Hy=3`:

image

**The fold is correct if all halos are filled and the circles are centered around the stars (= the pivot points).**

---

For a 4x2 FPivot:

image

---

And I also checked the halos for each rank, e.g., 4x2 UPivot rank 7:

image

---

For the record, it currently looks like that for the main branch, missing top halo row and fold line for example + some incorrectly pivoted cells (non-concentric circles):

image

---

So overall I think that the src/ part is ready for review. 

---

Follows from #5408.
Supersedes #5381.

### CODE REVIEW

File: src/OrthogonalSphericalShellGrids/tripolar_grid.jl [hunk 1]
```diff
@@ -1,24 +1,37 @@
 using Oceananigans.BoundaryConditions: UPivotZipperBoundaryCondition, FPivotZipperBoundaryCondition, NoFluxBoundaryCondition
 using Oceananigans.Grids: Grids, Bounded, Flat, OrthogonalSphericalShellGrid, Periodic, RectilinearGrid,
     architecture, cpu_face_constructor_z, validate_dimension_specification,
-    RightCenterFolded, RightFaceFolded
+    AbstractTopology, RightCenterFolded, RightFaceFolded
 using Oceananigans.ImmersedBoundaries: ImmersedBoundaryGrid
 
 """
-    struct Tripolar{N, F, S}
+    struct Tripolar{N, F, S, FT<:AbstractTopology}
```
Comment (simone-silvestri) [inline src/OrthogonalSphericalShellGrids/tripolar_grid.jl]:
```suggestion
    struct Tripolar{N, F, S, TY<:AbstractTopology}
```
`FT` is ambiguous, we typically use it for floating type

File: src/OrthogonalSphericalShellGrids/distributed_zipper.jl [hunk 2]
```diff
@@ -1,57 +1,471 @@
...
+        view(buff.recv, :, 2:size(buff.recv,2), :)
+
+_recv_from_northeast_buffer!(c, buff::ZipperCornerBuffer{false}, Hx, Hy, Nx, Ny) =
+    view(c, ne_recv_x(loc_x(buff), Hx, Nx), recv_halo_y(fold_topo(buff), loc_y(buff), Hy, Ny), :) .= buff.recv
+
+#####
+##### _switch_north_halos! — no-op (fold logic in send buffers)
+#####
+
+switch_north_halos!(c, north_bc, grid, loc) = nothing
```
Comment (simone-silvestri) [inline src/OrthogonalSphericalShellGrids/distributed_zipper.jl]:
```suggestion
```

File: src/OrthogonalSphericalShellGrids/distributed_zipper.jl [hunk 3]
```diff
@@ -68,40 +482,24 @@ function fill_halo_regions!(c::OffsetArray, bcs, indices, loc, grid::MPITripolar
...
 
-    if arch.mpi_tag[] == 0 # The communication has been reset, switch the north halos!
-        north_bc = bcs.north
-        switch_north_halos!(c, north_bc, grid, loc)
-    end
-
     return nothing
 end
 
 function synchronize_communication!(field::Field{<:Any, <:Any, <:Any, <:Any, <:MPITripolarGridOfSomeKind})
```
Comment (simone-silvestri) [inline src/OrthogonalSphericalShellGrids/distributed_zipper.jl]: this can probably be deleted

File: src/OrthogonalSphericalShellGrids/distributed_zipper.jl [hunk 4]
```diff
@@ -1,57 +1,471 @@
...
+
+switch_north_halos!(c, north_bc, grid, loc) = nothing
+
+#####
+##### fill_halo_regions! for distributed tripolar grids
+#####
+
 fill_halo_regions!(c::OffsetArray, ::Nothing, indices, loc, ::MPITripolarGridOfSomeKind, args...; kwargs...) = nothing
 
 function fill_halo_regions!(c::OffsetArray, bcs, indices, loc, grid::MPITripolarGridOfSomeKind, buffers::CommunicationBuffers, args...; kwargs...)
```
Comment (simone-silvestri) [inline src/OrthogonalSphericalShellGrids/distributed_zipper.jl]: most likely these can be deleted if the logic is in the fill and recv buffers

File: src/OrthogonalSphericalShellGrids/tripolar_grid.jl [hunk 5]
```diff
@@ -1,24 +1,37 @@
...
 """
-    struct Tripolar{N, F, S}
+    struct Tripolar{N, F, S, TY<:AbstractTopology}
 
 A structure to represent a tripolar grid on an orthogonal spherical shell.
+The fold topology `FT` (e.g., `RightCenterFolded` or `RightFaceFolded`) is stored
+as a type parameter rather than a field, keeping the struct `isbits` for GPU kernels.
 """
-struct Tripolar{N, F, S}
+struct Tripolar{N, F, S, FT<:AbstractTopology}
```
Comment (simone-silvestri) [inline src/OrthogonalSphericalShellGrids/tripolar_grid.jl]:
```suggestion
struct Tripolar{N, F, S, TY<:AbstractTopology}
```

File: src/OrthogonalSphericalShellGrids/tripolar_grid.jl [hunk 6]
```diff
@@ -1,24 +1,37 @@
...
     north_poles_latitude :: N
     first_pole_longitude :: F
     southernmost_latitude :: S
 end
 
-Adapt.adapt_structure(to, t::Tripolar) =
+# Getter: returns the fold topology Type
+fold_topology(::Tripolar{<:Any, <:Any, <:Any, FT}) where FT = FT
```
Comment (simone-silvestri) [inline src/OrthogonalSphericalShellGrids/tripolar_grid.jl]:
```suggestion
fold_topology(::Tripolar{<:Any, <:Any, <:Any, TY}) where TY = TY
```

File: src/OrthogonalSphericalShellGrids/tripolar_grid.jl [hunk 7]
```diff
@@ -1,24 +1,37 @@
...
+
+# Constructor accepting fold topology as a Type argument
+Tripolar(n, f, s, ::Type{FT}) where {FT<:AbstractTopology} =
+    Tripolar{typeof(n), typeof(f), typeof(s), FT}(n, f, s)
```
Comment (simone-silvestri) [inline src/OrthogonalSphericalShellGrids/tripolar_grid.jl]:
```suggestion
Tripolar(n, f, s, ::Type{TY}) where {TY<:AbstractTopology} =
    Tripolar{typeof(n), typeof(f), typeof(s), TY}(n, f, s)
```

File: src/OrthogonalSphericalShellGrids/tripolar_grid.jl [hunk 8]
```diff
@@ -1,24 +1,37 @@
...
+
+# Backward-compatible constructor (defaults to UPivot)
+Tripolar(n, f, s) = Tripolar(n, f, s, RightCenterFolded)
+
+Adapt.adapt_structure(to, t::Tripolar{<:Any, <:Any, <:Any, FT}) where FT =
     Tripolar(Adapt.adapt(to, t.north_poles_latitude),
              Adapt.adapt(to, t.first_pole_longitude),
-             Adapt.adapt(to, t.southernmost_latitude))
+             Adapt.adapt(to, t.southernmost_latitude),
+             FT)
```
Comment (simone-silvestri) [inline src/OrthogonalSphericalShellGrids/tripolar_grid.jl]:
```suggestion
Adapt.adapt_structure(to, t::Tripolar{<:Any, <:Any, <:Any, TY}) where TY =
    Tripolar(Adapt.adapt(to, t.north_poles_latitude),
             Adapt.adapt(to, t.first_pole_longitude),
             Adapt.adapt(to, t.southernmost_latitude),
             TY)
```

File: validation/orthogonal_spherical_shell_grid/halo_fill_viz.jl [hunk 9]
```diff
@@ -60,7 +60,7 @@ for fold_topology in fold_topologies
         color = :red, marker = :star5, markersize = 15, label = "Pivot points"
         )
         # default grid underlay
-        offset = (fold_topology == RightFaceFolded) ? 1 : 1/2 # how much more grid north of pivots
+        offset = (fold_topology == RightFaceFolded) ? 0 : 1/2 # how much more grid north of pivots
```
Comment (simone-silvestri) [inline validation/orthogonal_spherical_shell_grid/halo_fill_viz.jl:63]: is this different now?

Comment (briochemc) [reply validation/orthogonal_spherical_shell_grid/halo_fill_viz.jl:63]: Yes this just the updated validation script for #5408 (we forgot to update it then).

File: src/OrthogonalSphericalShellGrids/tripolar_grid.jl (hunk 8 above)
Comment (briochemc) [reply src/OrthogonalSphericalShellGrids/tripolar_grid.jl]: Can't seem to apply it from GitHub. Will do it locally.

File: src/OrthogonalSphericalShellGrids/distributed_zipper.jl (hunk 3 above)
Comment (briochemc) [reply src/OrthogonalSphericalShellGrids/distributed_zipper.jl]: Done in https://github.com/CliMA/Oceananigans.jl/pull/5439/commits/7695d8a7bf62420b22c2c6ceeed3fcabcb7821b2

File: src/OrthogonalSphericalShellGrids/distributed_zipper.jl (hunk 4 above)
Comment (briochemc) [reply src/OrthogonalSphericalShellGrids/distributed_zipper.jl]: Done in https://github.com/CliMA/Oceananigans.jl/pull/5439/commits/7695d8a7bf62420b22c2c6ceeed3fcabcb7821b2

**Comment by briochemc** [comment]
I have added tests for `RightFaceFolded` (`FPivot`), which passed locally. I also profiled the PR against main on 4 GPUs (1x4 and 2x2) and it was performing similar on 1x4 and better on 2x2 (cc @simone-silvestri, with whom I shared some Nsight screenshots). For me this is now ready for review/merge.  

The only thing I haven't done is add my extra tests: the index-tracing and the per-rank halo comparison with serial. I'm worried that if I add these "on top" of the current tests, it will be too slow to finish (given that I roughly doubled these tests by adding the extra fold topology). Should I add another script instead so that it can run in parallel to the existing tests?

File: src/BoundaryConditions/field_boundary_conditions.jl [hunk 10]
```diff
@@ -21,13 +21,18 @@ default_prognostic_bc(::RightConnected, ::Center, default)  = default.boundary_c
...
+
+default_prognostic_bc(::DistributedFoldTopology, ::Center, default) = default.boundary_condition
 
 # TODO: make model constructors enforce impenetrability on velocity components to simplify this code
 default_prognostic_bc(::Bounded,        ::Face, default) = ImpenetrableBoundaryCondition()
 default_prognostic_bc(::LeftConnected,  ::Face, default) = ImpenetrableBoundaryCondition()
 default_prognostic_bc(::RightConnected, ::Face, default) = ImpenetrableBoundaryCondition()
 default_prognostic_bc(::RightFaceFolded, ::Face, default) = ImpenetrableBoundaryCondition()
 default_prognostic_bc(::RightCenterFolded, ::Face, default) = ImpenetrableBoundaryCondition()
+default_prognostic_bc(::DistributedFoldTopology, ::Face, default) = ImpenetrableBoundaryCondition()
```
Comment (simone-silvestri) [inline src/BoundaryConditions/field_boundary_conditions.jl:35]: wouldn't this be some pivot BC?

File: src/DistributedComputations/distributed_grids.jl [hunk 11]
```diff
@@ -349,6 +349,26 @@ insert_connected_topology(::Type{Bounded}, R, r) = ifelse(R == 1, Bounded,
...
+# These take 5 arguments: (global_y_topology, Ry, ry, Rx, rx) where
+# Ry/ry are y-rank count/index and Rx/rx are x-rank count/index (all 1-based).
+
+local_fold_topology(::Type{RightCenterFolded}, Rx, rx) =
+    Rx == 1 ? LeftConnectedRightCenterFolded : LeftConnectedRightCenterConnected
+
+local_fold_topology(::Type{RightFaceFolded}, Rx, rx) =
+    Rx == 1 ? LeftConnectedRightFaceFolded : LeftConnectedRightFaceConnected
+
+function insert_connected_topology(T::Type{<:Union{RightCenterFolded, RightFaceFolded}}, Ry, ry, Rx, rx)
```
Comment (simone-silvestri) [inline src/DistributedComputations/distributed_grids.jl:362]: is this used somewhere? the signature is not the same as for the other methods

Comment (simone-silvestri) [reply src/DistributedComputations/distributed_grids.jl:362]: ah ok I see, it is used later in the OrthogonalSphericalShellGrids module, maybe we can move it there

File: src/OrthogonalSphericalShellGrids/distributed_tripolar_grid.jl [hunk 12]
```diff
@@ -109,8 +112,14 @@ function TripolarGrid(arch::Distributed, FT::DataType=Float64;
...
     LX = workers[1] == 1 ? Periodic : FullyConnected
+
+    global_fold_topology = topology(global_grid, 2)
+
+    # 1-based indices for insert_connected_topology
+    Rx, Ry = workers[1], workers[2]
+    rx, ry = xrank + 1, yrank + 1
+    LY = insert_connected_topology(global_fold_topology, Ry, ry, Rx, rx)
     ny = nylocal[yrank+1]
     nx = nxlocal[xrank+1]
```
Comment (simone-silvestri) [inline src/OrthogonalSphericalShellGrids/distributed_tripolar_grid.jl]:
```suggestion
    ny = nylocal[ry]
    nx = nxlocal[rx]
```

File: src/OrthogonalSphericalShellGrids/distributed_tripolar_grid.jl [hunk 13]
```diff
@@ -248,7 +254,7 @@ function regularize_field_boundary_conditions(bcs::FieldBoundaryConditions,
     south = regularize_boundary_condition(bcs.south, grid, loc, 2, LeftBoundary,  prognostic_names)
 
     north = if yrank == processor_size[2] - 1 && processor_size[1] == 1
-        UPivotZipperBoundaryCondition(sign)
+        north_fold_boundary_condition(fold_topology(grid.conformal_mapping))(sign)
```
Comment (simone-silvestri) [inline src/OrthogonalSphericalShellGrids/distributed_tripolar_grid.jl]:
```suggestion
        TY = fold_topology(grid.conformal_mapping)
        north_fold_boundary_condition(TY)(sign)
```

File: src/BoundaryConditions/field_boundary_conditions.jl (hunk 10 above)
Comment (briochemc) [reply src/BoundaryConditions/field_boundary_conditions.jl:35]: I was curious about that but I opted to follow the existing pattern for `RightFaceFolded` and `RightCenterFolded`. Should I change that?

File: src/DistributedComputations/distributed_grids.jl (hunk 11 above)
Comment (briochemc) [reply src/DistributedComputations/distributed_grids.jl:362]: Ah yes sure, I thought I'd put it next to what it extends but no worries :)

File: src/BoundaryConditions/field_boundary_conditions.jl (hunk 10 above)
Comment (simone-silvestri) [reply src/BoundaryConditions/field_boundary_conditions.jl:35]: maybe in another PR

File: src/OrthogonalSphericalShellGrids/distributed_tripolar_grid.jl [hunk 14]
```diff
@@ -280,26 +286,20 @@ function Field(loc::Tuple{<:LX, <:LY, <:LZ}, grid::MPITripolarGridOfSomeKind, da
...
 
-        # North boundary conditions are "special". If we are at the top of the domain, i.e.
-        # the last rank, then we need to substitute the BC only if the old one is not already
-        # a zipper boundary condition. Otherwise we always substitute because we need to
-        # inject the halo boundary conditions.
         if yrank == processor_size[2] - 1 && processor_size[1] == 1
-            north_bc = if !(old_bcs.north isa UZBC)
-                default_zipper
+            north_bc = if !(old_bcs.north isa ZBC)
+                north_fold_boundary_condition(fold_topology(grid.conformal_mapping))(sign(LX, LY))
```
Comment (simone-silvestri) [inline src/OrthogonalSphericalShellGrids/distributed_tripolar_grid.jl]: this line is a bit cluttered

Comment (briochemc) [reply src/OrthogonalSphericalShellGrids/distributed_tripolar_grid.jl]: I'll apply the same change you suggested just above.

File: src/OrthogonalSphericalShellGrids/distributed_zipper.jl [hunk 15]
```diff
@@ -1,108 +1,408 @@
...
+    rx = arch.local_index[1]
+    Rx = ranks(arch)[1]
+    return rx >= Rx ÷ 2 && rx < Rx
+end
+
+#####
+##### Zipper communication buffers with fold-aware packing
+#####
+
+struct TwoDZipperBuffer{FL, WFL, Loc, FoT, B, S}
```
Comment (simone-silvestri) [inline src/OrthogonalSphericalShellGrids/distributed_zipper.jl]: is FoT the fold topology? If yes can we use TY throughout?

File: src/OrthogonalSphericalShellGrids/distributed_tripolar_grid.jl (hunk 14 above)
Comment (briochemc) [reply src/OrthogonalSphericalShellGrids/distributed_tripolar_grid.jl]: fixed in ca61b8e47a06388194ee88f08afe89f06f5f6d50.

File: src/OrthogonalSphericalShellGrids/distributed_zipper.jl (hunk 15 above)
Comment (briochemc) [reply src/OrthogonalSphericalShellGrids/distributed_zipper.jl]: Yes no problem

File: src/OrthogonalSphericalShellGrids/distributed_zipper.jl [hunk 16]
```diff
@@ -1,108 +1,408 @@
...
-    adjust_x_face!(c, loc, north_halos, size(c, 1))
+# TripolarXBuffer recv FL=true, WFL=false: skip last row (the fold line)
+_recv_from_west_buffer!(c, buff::TripolarXBuffer{<:Any, true, false}, Hx, Hy, Nx, Ny) =
+    view(c, 1:Hx, 1+Hy:size(buff.recv,2)-1+Hy, :) .= view(buff.recv, :, 1:size(buff.recv,2)-1, :)
+_recv_from_east_buffer!(c, buff::TripolarXBuffer{<:Any, true, false}, Hx, Hy, Nx, Ny) =
+    view(c, 1+Nx+Hx:Nx+2Hx, 1+Hy:size(buff.recv,2)-1+Hy, :) .= view(buff.recv, :, 1:size(buff.recv,2)-1, :)
 
-    return nothing
+# Fold-aware x-buffer constructors.
+# Fallback for non-zipper north (south ranks): standard x-communication.
```
Comment (simone-silvestri) [inline src/OrthogonalSphericalShellGrids/distributed_zipper.jl]: Is this the same buffer as the standard one?

File: src/OrthogonalSphericalShellGrids/distributed_zipper.jl [hunk 17]
```diff
@@ -1,108 +1,408 @@
...
+#####
 
-    for task = 1:number_of_tasks
-        @inbounds distributed_fill_halo_event!(c, kernels![task], ordered_bcs[task], loc, grid, buffers, args...; kwargs...)
-    end
+function _fill_north_send_buffer!(c, b::TwoDZipperBuffer{true}, Hx, Hy, Nx, Ny)
+    topo, ly = fold_topo(b), loc_y(b)
+    fy = send_fold_y(topo, ly, Hy, Ny)
+    view(b.send, :, 1:1, :)    .= b.sign .* view(c, Nx+Hx:-1:1+Hx, fy:fy, :)
+    view(b.send, :, 2:Hy+1, :) .= b.sign .* view(c, Nx+Hx:-1:1+Hx, send_halo_y(topo, ly, Hy, Ny), :)
```
Comment (simone-silvestri) [inline src/OrthogonalSphericalShellGrids/distributed_zipper.jl]: this line also could be decluttered a bit

File: src/OrthogonalSphericalShellGrids/distributed_zipper.jl [hunk 18]
```diff
@@ -1,108 +1,408 @@
...
+end
 
-    fill_corners!(c, arch.connectivity, indices, loc, arch, grid, buffers, args...; kwargs...)
+function _fill_north_send_buffer!(c, b::TwoDZipperBuffer{false}, Hx, Hy, Nx, Ny)
+    topo, ly = fold_topo(b), loc_y(b)
+    b.send .= b.sign .* view(c, Nx+Hx:-1:1+Hx, send_halo_y(topo, ly, Hy, Ny), :)
```
Comment (simone-silvestri) [inline src/OrthogonalSphericalShellGrids/distributed_zipper.jl]: and here also

File: src/OrthogonalSphericalShellGrids/distributed_zipper.jl (hunk 15 above)
Comment (briochemc) [reply src/OrthogonalSphericalShellGrids/distributed_zipper.jl]: Done in 0d98e06684b66d4ae70a1d972b575ca3c83616fc

File: src/OrthogonalSphericalShellGrids/distributed_zipper.jl (hunk 16 above)
Comment (briochemc) [reply src/OrthogonalSphericalShellGrids/distributed_zipper.jl]: Yes, although I think this comment should not be there. There is already a comment for it L169.

Comment (briochemc) [reply src/OrthogonalSphericalShellGrids/distributed_zipper.jl]: Done in d4d24326222f73a04296f8c9304129b772ee0a07

File: src/OrthogonalSphericalShellGrids/distributed_zipper.jl (hunk 17 above)
Comment (briochemc) [reply src/OrthogonalSphericalShellGrids/distributed_zipper.jl]: Done in 70a4d2b79f9c1f0abf0b2b9c76ff546e01a76d20

File: src/OrthogonalSphericalShellGrids/distributed_zipper.jl (hunk 18 above)
Comment (briochemc) [reply src/OrthogonalSphericalShellGrids/distributed_zipper.jl]: Done in 70a4d2b79f9c1f0abf0b2b9c76ff546e01a76d20

**Comment by briochemc** [comment]
> I would declutter a bit all the code in the `distributed_zipper.jl` file. For example lines like
> 
> ```julia
> _recv_from_northwest_buffer!(c, buff::ZipperCornerBuffer{false}, Hx, Hy, Nx, Ny) =
>     view(c, nw_recv_x(loc_x(buff), Hx, Nx), recv_halo_y(fold_topo(buff), loc_y(buff), Hy, Ny), :) .= buff.recv
> ```
> 
> end up being very difficult to read. A couple of lines more facilitate the reading

Agreed. Made a bunch of changes for this in 70a4d2b79f9c1f0abf0b2b9c76ff546e01a76d20.

File: src/OrthogonalSphericalShellGrids/distributed_zipper.jl (hunk 18 above)
Comment (briochemc) [reply src/OrthogonalSphericalShellGrids/distributed_zipper.jl]: Done in 70a4d2b79f9c1f0abf0b2b9c76ff546e01a76d20

File: src/OrthogonalSphericalShellGrids/distributed_zipper.jl (hunk 17 above)
Comment (briochemc) [reply src/OrthogonalSphericalShellGrids/distributed_zipper.jl]: Done in 70a4d2b79f9c1f0abf0b2b9c76ff546e01a76d20

File: src/DistributedComputations/distributed_grids.jl (hunk 11 above)
Comment (briochemc) [reply src/DistributedComputations/distributed_grids.jl:362]: Where should it go in the module? In distributed_tripolar_grid.jl?

File: src/OrthogonalSphericalShellGrids/distributed_zipper.jl [hunk 19]
```diff
@@ -208,7 +207,7 @@ function y_tripolar_buffer(arch, grid::AbstractGrid{<:Any, <:Any, Topo},
     wfl = fl && north_writes_fold_line(arch)
     send = on_architecture(arch, zeros(FT, Nx, Hy′, Tz))
     recv = on_architecture(arch, zeros(FT, Nx, Hy′, Tz))
-    return TwoDZipperBuffer(loc, Topo(), send, recv, sgn, Val(fl), Val(wfl))
+    return TwoDZipperBuffer{fl, wfl, Topo}(loc, send, recv, sgn)
```
Comment (simone-silvestri) [inline src/OrthogonalSphericalShellGrids/distributed_zipper.jl]: can Topo be TY?

File: src/OrthogonalSphericalShellGrids/distributed_zipper.jl [hunk 20]
```diff
@@ -221,29 +220,29 @@ northeast_tripolar_buffer(arch, grid, data, Hx, Hy, xedge, yedge) = corner_commu
 # based on whether the corner's global x-position is past the pivot.
 
 function northwest_tripolar_buffer(arch, grid, data, Hx, Hy, xedge,
-                                   yedge::TwoDZipperBuffer{<:Any, <:Any, Loc, TY}) where {Loc, TY}
+                                   yedge::TwoDZipperBuffer{<:Any, <:Any, TY}) where TY
     Tz = size(parent(data), 3); FT = eltype(data); sgn = yedge.sign
-    loc_x_inst = instantiate(Loc.parameters[1])
-    fl = has_fold_line(TY, Loc.parameters[2]())
+    lx, ly = yedge.loc[1], yedge.loc[2]
```
Comment (simone-silvestri) [inline src/OrthogonalSphericalShellGrids/distributed_zipper.jl]: In other places for instantiated locations we use `\ell`

File: src/OrthogonalSphericalShellGrids/distributed_zipper.jl [hunk 21]
```diff
@@ -259,87 +258,87 @@ const FF = Tuple{<:Face,   <:Face,   <:Any}
...
 #####
 ##### TwoDZipperBuffer: north send (FL=true has fold line, FL=false does not)
 #####
 
 function _fill_north_send_buffer!(c, b::TwoDZipperBuffer{true}, Hx, Hy, Nx, Ny)
-    topo, ly = fold_topo(b), loc_y(b)
-    fy = send_fold_y(topo, ly, Hy, Ny)
-    view(b.send, :, 1:1, :)    .= b.sign .* view(c, Nx+Hx:-1:1+Hx, fy:fy, :)
-    view(b.send, :, 2:Hy+1, :) .= b.sign .* view(c, Nx+Hx:-1:1+Hx, send_halo_y(topo, ly, Hy, Ny), :)
+    topo, ly = fold_topo(b), b.loc[2]
```
Comment (simone-silvestri) [inline src/OrthogonalSphericalShellGrids/distributed_zipper.jl]: is this TY?

File: src/OrthogonalSphericalShellGrids/distributed_zipper.jl (hunk 19 above)
Comment (briochemc) [reply src/OrthogonalSphericalShellGrids/distributed_zipper.jl]: Done.

File: src/OrthogonalSphericalShellGrids/distributed_zipper.jl (hunk 20 above)
Comment (briochemc) [reply src/OrthogonalSphericalShellGrids/distributed_zipper.jl]: Done

File: src/OrthogonalSphericalShellGrids/distributed_zipper.jl (hunk 21 above)
Comment (briochemc) [reply src/OrthogonalSphericalShellGrids/distributed_zipper.jl]: No but it's `TY()` (instantiated)

**Comment by simone-silvestri** [comment]
LGTM
